### ISSUE TITLE
Add support to timestamp resolution other than microsecond for TWCS

### ISSUE THREAD

**Issue by raphaelsc** [scylladb/scylla#3152]
Certain/older clients may decide to work with milliseconds as timestamp resolution instead

**Comment by slivne** [comment]
@raphaelsc can you please clarify the issue with different clients ? 

if we have two clients writing data with different resolution how will the timestamp info be persisted into the file ? 

are other compaction strategies resilient to this - and if so how ?

**Comment by raphaelsc** [comment]
On Sun, Jan 28, 2018 at 9:20 AM, Shlomi Livne <notifications@github.com>
wrote:

> @raphaelsc <https://github.com/raphaelsc> can you please clarify the
> issue with different clients ?
>
> if we have two clients writing data with different resolution how will the
> timestamp info be persisted into the file ?
>

AFAICT, user supplied timestamp will be persisted as is, so user is
responsible to keeping timestamp unit consistent for a given column family.
Perhaps we check that a timestamp has millisecond resolution and convert it
into micro before forwarding to sstable write? I don't think so. I need to
double check this.


> are other compaction strategies resilient to this - and if so how ?
>

I don't think either TWCS or DTCS will be resilient to this. They have the
timestamp resolution option, defined in schema, and expects timestamp in
sstable metadata (min, max) to be exactly of the unit specified in schema.

from datastax doc: "When setting the timestamp_resolution for Date Tiered
Compaction Strategy (DTCS) settings, it is important to understand why
writes also need to set the same timestamp resolution. Failure to do so can
lead to a unmanageable number of sstables which can in turn lead to
compaction and repair problems."

so same requirement for C*.


What we need to do here is to allow an user willing to work with only
milliseconds to be able to do so, by extending TWCS to support it.
Currently, TWCS will only work with user supplied or server generated
timestamp which uses microsecond as resolution.

**Comment by slivne** [comment]
@raphaelsc well it seems strange that the user provided timestamp is not validated against the schema - I would expect some validation to be done.

**Comment by raphaelsc** [comment]
On Sun, Feb 4, 2018 at 10:08 AM, Shlomi Livne <notifications@github.com>
wrote:

> @raphaelsc <https://github.com/raphaelsc> well it seems strange that the
> user provided timestamp is not validated against the schema - I would
> expect some validation to be done.
>

It's probably a bad design issue. The timestamp resolution option exists
only for time series strategies, so it'd have to be done at strategy level.
So we'd need to postpone validation to strategy when extracting timestamp
from sstables.

We could try to add some resilience by converting ts on the fly, or throw
exception which would result in shutdown. Either way is better than current
because strategy will silently work with whatever timestamp it's provided
with which can lead to it misbehaving and consequently user will have bad
read performance.

**Comment by tzach** [comment]
> Certain/older clients may decide to work with milliseconds as timestamp resolution instead

which client/versions?
In particular, Kairos 1.2.0 is using cassandra-driver-core 3.3.2[1]
Is this driver version problematic?

[1] https://github.com/kairosdb/kairosdb/blob/v1.2.0/ivy.xml#L61

**Comment by gcmcnutt** [comment]
we too are blocked on using TWCS with kairosDB -- if anyone can come up with a workaround or hackaround, even in kairosdb, in the mean time, that would help

**Comment by avikivity** [comment]
Is this about the `compaction_window_unit` of TWCS? If so, I don't think it affects timestamp resolution.

If not, what is it?

**Comment by raphaelsc** [comment]
On Sun, Apr 8, 2018 at 8:58 AM, Avi Kivity <notifications@github.com> wrote:

> Is this about the compaction_window_unit of TWCS? If so, I don't think it
> affects timestamp resolution.
>
> If not, what is it?
>
the timestamp resolution provided by client, which goes in
timestamp_resolution option which in turn goes in compaction strategy
options. That's somewhat awful. Older clients use millisecond instead of
microsecond.

**Comment by raphaelsc** [comment]
Patch sent, waiting for review...

**Comment by avikivity** [comment]
Wow, that's truly terrible. Oh well.

**Comment by dorlaor** [comment]
It will be nice to push to 2.2

On Thu, Apr 12, 2018 at 3:35 AM, scylladb-promoter <notifications@github.com
